cdap-ui/app/features/hydratorplusplus/controllers/create/toppanel-ctrl.js

```diff
@@ -92,6 +124,116 @@ class HydratorPlusPlusTopPanelCtrl{
     }
   }
 
+  startTimer() {
+    this.previewTimerInterval = this.$interval(() => {
+      let duration = (new Date() - this.previewStartTime) / 1000;
+      duration = duration >= 0 ? duration : 0;
+
+      let minutes = Math.floor(duration / 60);
+      let seconds = Math.floor(duration % 60);
+      seconds = seconds < 10 ? '0' + seconds : seconds;
+      minutes = minutes < 10 ? '0' + minutes : minutes;
+
+      this.displayDuration = {
+        minutes: minutes,
+        seconds: seconds
+      };
+    }, 500);
+  }
+  stopTimer() {
+    this.$interval.cancel(this.previewTimerInterval);
+  }
+
+  runPreview() {
+    let startTime = new Date();
+    this.previewStartTime = startTime;
+    this.startTimer();
+    this.previewStore.dispatch(
+      this.previewActions.setPreviewStartTime(startTime)
+    );
+
+    let params = {
+      namespace: this.$state.params.namespace,
+      scope: this.$scope
+    };
+
+
+    // GENERATING PREVIEW CONFIG
+    // This might/should be extracted out to a factory
+
+    let pipelineConfig = this.HydratorPlusPlusConfigStore.getConfigForExport();
+    /**
+     *  This is a cheat way for generating preview for the entire pipeline
+     **/
+
+    let previewConfig = {
+      startStages: [],
+      endStages: [],
+      // useSinks: [], // we are not using sinks for now
+      numOfRecords: 10
+    };
+
+    // Get start stages and end stages
+    // Current implementation:
+    //    - start stages mean sources
+    //    - end stages mean sinks
+    angular.forEach(pipelineConfig.config.stages, (node) => {
+      if (this.GLOBALS.pluginConvert[node.plugin.type] === 'source') {
+        previewConfig.startStages.push(node.name);
+      } else if (this.GLOBALS.pluginConvert[node.plugin.type] === 'sink') {
+        previewConfig.endStages.push(node.name);
+      }
+    });
+    pipelineConfig.config.preview = previewConfig;
+
+    if (previewConfig.startStages.length === 0 || previewConfig.endStages.length === 0) {
```

Comment:
I'm not sure at this point. I think we should be able to reuse the normal error factory for validation